File: core/parser/throne.py
"""
Парсер конфигураций формата Throne (ссылки throne:// и JSON).
"""

import json
import urllib.parse
from typing import Optional

from core.config.models import ProfileConfig, ProtocolType
from core.parser.base import BaseParser, ParseResult
from utils.logger import log_call


class ThroneParser(BaseParser):
    """Парсер для формата Throne."""

    format_name: str = "throne"

    @log_call
    def parse(self, content: str) -> ParseResult:
        """
        Разбирает ссылки формата throne:// или JSON Throne.

        Args:
            content: Текст ссылки или конфигурации.

        Returns:
            ParseResult: Результат парсинга.
        """
        content = content.strip()
        result = ParseResult(format_name=self.format_name)

        if content.startswith("{") and content.endswith("}"):
            try:
                data = json.loads(content)
                result.raw_config = data
                servers = data.get("servers", [])
                for srv in servers:
                    if isinstance(srv, dict):
                        result.profiles.append(
                            ProfileConfig(
                                name=srv.get("name", "Throne Proxy"),
                                protocol=ProtocolType(srv.get("protocol", "vless")),
                                server=srv.get("server", "127.0.0.1"),
                                server_port=int(srv.get("port", 443)),
                                settings=srv.get("params", {}),
                            )
                        )
                return result
            except json.JSONDecodeError as exc:
                result.errors.append(f"Ошибка парсинга Throne JSON: {exc}")
                return result

        lines = [line.strip() for line in content.splitlines() if line.strip()]
        for line in lines:
            if line.startswith("throne://"):
                profile = self._parse_throne_link(line)
                if profile:
                    result.profiles.append(profile)

        return result

    @log_call
    def _parse_throne_link(self, uri: str) -> Optional[ProfileConfig]:
        """Парсит отдельную throne:// ссылку."""
        parsed = urllib.parse.urlparse(uri)
        name = urllib.parse.unquote(parsed.fragment) or "Throne Node"
        params = urllib.parse.parse_qs(parsed.query)

        server = parsed.hostname or "127.0.0.1"
        port = parsed.port or 443

        protocol_str = params.get("mode", ["vless"])[0]
        try:
            protocol = ProtocolType(protocol_str)
        except ValueError:
            protocol = ProtocolType.VLESS

        return ProfileConfig(
            name=name,
            protocol=protocol,
            server=server,
            server_port=port,
            settings={k: v[0] for k, v in params.items()},
        )

File: core_manager/src/parser/throne.rs
//! Парсер Throne конфигураций на Rust.

use crate::models::{ProfileConfig, ProtocolType};
use crate::parser::{ConfigParser, ParseResult};

pub struct ThroneParser;

impl ConfigParser for ThroneParser {
    fn format_name(&self) -> &'static str {
        "throne"
    }

    fn parse(&self, content: &str) -> ParseResult {
        let text = content.trim();
        let mut result = ParseResult::new(self.format_name());

        if text.starts_with('{') && text.ends_with('}') {
            if let Ok(json_val) = serde_json::from_str::<serde_json::Value>(text) {
                result.raw_config = Some(json_val.clone());
                if let Some(servers) = json_val.get("servers").and_then(|s| s.as_array()) {
                    for item in servers {
                        let name = item.get("name").and_then(|n| n.as_str()).unwrap_or("Throne Proxy");
                        let server = item.get("server").and_then(|s| s.as_str()).unwrap_or("127.0.0.1");
                        let port = item.get("port").and_then(|p| p.as_u64()).unwrap_or(443) as u16;

                        let p = ProfileConfig::new(name, ProtocolType::Vless, server, port);
                        result.profiles.push(p);
                    }
                    return result;
                }
            }
        }

        for line in text.lines() {
            let l = line.trim();
            if l.starts_with("throne://") {
                if let Ok(url) = url::Url::parse(l) {
                    let host = url.host_str().unwrap_or("127.0.0.1");
                    let port = url.port().unwrap_or(443);
                    let name = url.fragment().unwrap_or("Throne Node");

                    let p = ProfileConfig::new(name, ProtocolType::Vless, host, port);
                    result.profiles.push(p);
                }
            }
        }

        result
    }
}
